### PR TITLE
utility providers page and user payment

### PR DESCRIPTION
**Pull request by Andrii-Kozak**
[Trello Task](https://trello.com/_____)

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs
- [x] screenshot of feature is added if task has UI

image
image

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -55,12 +57,25 @@ def self.grouped_collection_by_role
     }
   end
 
+
   def companies_for_output
     companies.map(&:name).join(', ')
+
+  def balance_total
+    payments_total - bills_total
+
```
Comment (pereslop) [inline app/models/user.rb:66]: remove empty line

File: app/views/account/utility_providers/_utility_provider.html.slim [hunk 2]
```diff
@@ -1,20 +1,39 @@
...
+              = I18n.t("account.utility_providers.utility_provider.amount",
+                amount: utility_provider.balance_utility_provider.ceil(2))
+          - else utility_provider.balance_utility_provider < 0
+            .text-danger
+              = I18n.t("account.utility_providers.utility_provider.amount",
+                amount: utility_provider.balance_utility_provider.ceil(2))
+    .col-md-3.btn-pay
+      - if utility_provider.balance_utility_provider < 0
+        button.btn.btn-outline-primary.btn-pay-custom
+          |Oплата
```
Comment (pereslop) [inline app/views/account/utility_providers/_utility_provider.html.slim:37]: I18n

File: app/models/user.rb (hunk 1 above)
Comment (mmarusyk) [inline app/models/user.rb:66]: ???

File: app/views/account/utility_providers/index.html.slim [hunk 3]
```diff
@@ -1,28 +1,31 @@
...
+            = t('.last_payment')
+            = current_user.payments.last.updated_at.strftime("%d-%m-%Y")
+          - else
+            = t('.last_payment')
+        .text-right
+          = t('.total_payment')
+          - if current_user.balance_total >= 0
+            .text-success.d-inline-block = I18n.t("account.utility_providers.utility_provider.amount",
+              amount: current_user.balance_total.ceil(3))
+          - else current_user.balance_total < 0
```
Comment (pereslop) [inline app/views/account/utility_providers/index.html.slim:20]: else should be enough

File: app/views/account/utility_providers/show.html.slim [hunk 4]
```diff
@@ -1,24 +1,7 @@
...
-        class: 'btn btn-primary btn-custom'
-      = link_to t('.btn-delete'), disassociate_account_utility_provider_path(@billing_contract),
-        method: :put, data: { confirm: t('.are_you_sure') },
-        class: 'btn btn-danger btn-custom'
-      = link_to t('.btn-add'), new_account_utility_provider_meter_reading_path(utility_provider_id: @billing_contract.id),
-        class: 'btn btn-success btn-custom'
-      = link_to t('.btn-back'), account_utility_providers_path,
-        class: 'btn btn-dark btn-custom'
+h1
+  |show page
```
Comment (mmarusyk) [inline app/views/account/utility_providers/show.html.slim:2]: I18n

File: app/views/account/utility_providers/_utility_provider.html.slim [hunk 5]
```diff
@@ -1,20 +1,39 @@
...
+          =< utility_provider.contract_num
+      .row.row-content
+        .col-md.utility-text
+          = t('.balance')
+        .col-md.utility-text
+          - if utility_provider.balance_utility_provider >= 0
+            .text-success
+              = I18n.t("account.utility_providers.utility_provider.amount",
+                amount: utility_provider.balance_utility_provider.ceil(2))
+          - else utility_provider.balance_utility_provider < 0
```
Comment (pereslop) [inline app/views/account/utility_providers/_utility_provider.html.slim:30]: `utility_provider.balance_utility_provider < 0` you dont need it, only `else` keyword should be enought

File: app/javascript/stylesheets/osbbs.scss [hunk 6]
```diff
@@ -29,10 +29,6 @@
   align-items: center;
 }
 
-div a.btn-outline-primary {
-    margin-left: 30px;
-  }
-
```
Comment (marikrebega) [inline app/javascript/stylesheets/osbbs.scss:35]: Doesn't that change the styles for all the buttons? If so, make sure that these changes do not affect other pages, if not required

File: app/javascript/stylesheets/users.scss [hunk 7]
```diff
@@ -19,11 +19,8 @@
 }
 
 .date.optional select {
-  margin-right: 10px;
-}
-
-.date.optional label {
-  margin-right: 30px;
+  margin-right: 20px;
+  float: right;
```
Comment (marikrebega) [inline app/javascript/stylesheets/users.scss:23]: the same

File: db/seeds.rb [hunk 8]
```diff
@@ -39,8 +39,8 @@
 
 billing_contracts = BillingContract.all.shuffle
 
-User.select{ |user| user.role != 'admin' }.each do |user|
-  billing_contracts.pop(2).each do |billing_contract|
```
Comment (marikrebega) [inline db/seeds.rb:43]: Why did you delete it?

File: app/views/account/utility_providers/_utility_provider.html.slim (hunk 2 above)
Comment (Andrii-Kozak) [reply app/views/account/utility_providers/_utility_provider.html.slim]: fixed

File: app/models/user.rb (hunk 1 above)
Comment (Andrii-Kozak) [reply app/models/user.rb]: missed when resolving conflicts,
corrected)

File: app/javascript/stylesheets/osbbs.scss (hunk 6 above)
Comment (Andrii-Kozak) [reply app/javascript/stylesheets/osbbs.scss:35]: I checked it, everything is okay

File: app/javascript/stylesheets/users.scss (hunk 7 above)
Comment (Andrii-Kozak) [reply app/javascript/stylesheets/users.scss:23]: I checked it

File: db/seeds.rb (hunk 8 above)
Comment (Andrii-Kozak) [reply db/seeds.rb:43]: resolve)

File: app/views/account/utility_providers/_utility_provider.html.slim (hunk 5 above)
Comment (Andrii-Kozak) [reply app/views/account/utility_providers/_utility_provider.html.slim]: resolve)

File: app/views/account/utility_providers/index.html.slim (hunk 3 above)
Comment (Andrii-Kozak) [reply app/views/account/utility_providers/index.html.slim]: removed

File: app/models/user.rb (hunk 1 above)
Comment (Andrii-Kozak) [reply app/models/user.rb]: fixed

**Review by marikrebega — APPROVED** [review]

**Review by Swooty18 — APPROVED** [review]

File: app/javascript/stylesheets/utility_providers.scss [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+.row-pading {
+  padding: 0 15px;
+}
+
+.btn-pay-custom {
+  width: 80px;
+}
+@media(max-width: 1025px) {
```
Comment (mmarusyk) [inline app/javascript/stylesheets/utility_providers.scss:70]: Enter

File: app/models/billing_contract.rb [hunk 10]
```diff
@@ -10,6 +10,22 @@ class BillingContract < ApplicationRecord
...
+  end
+
+  private
+
+  def payments_contract
+    payments.sum(:amount).to_f
+  end
+
+  def bills_contract
+    bills.sum(:amount).to_f
```
Comment (mmarusyk) [inline app/models/billing_contract.rb:27]: Why do you use to_f?

**Review by matsevich — APPROVED** [review]

File: app/models/billing_contract.rb [hunk 11]
```diff
@@ -21,6 +23,15 @@ def self.to_csv
...
     end
   end
+
+  def last_bill
+    bills.ordered_by_date.last
+  end
+
+  def balance_utility_provider
+    payments.created_between(DateTime.now.beginning_of_month, Time.zone.now).sum(:amount) -
+      (last_bill.try(:amount) || 0)
```
Comment (chu4) [inline app/models/billing_contract.rb:33]: you have a validation that amount should be present

File: app/models/billing_contract.rb [hunk 12]
```diff
@@ -21,6 +23,15 @@ def self.to_csv
       end
     end
   end
+
+  def last_bill
```
Comment (chu4) [inline app/models/billing_contract.rb:27]: this should be a an association

File: app/models/billing_contract.rb [hunk 13]
```diff
@@ -21,6 +23,15 @@ def self.to_csv
...
+    bills.ordered_by_date.last
+  end
+
+  def balance_utility_provider
+    payments.created_between(DateTime.now.beginning_of_month, Time.zone.now).sum(:amount) -
```
Comment (chu4) [inline app/models/billing_contract.rb:32]: should not be DateTime.now.beginning_of_month but date of a bill

File: app/models/billing_contract.rb [hunk 14]
```diff
@@ -21,6 +23,15 @@ def self.to_csv
       end
     end
   end
+
+  def last_bill
+    bills.ordered_by_date.last
+  end
+
+  def balance_utility_provider
```
Comment (chu4) [inline app/models/billing_contract.rb:31]: method should receive bill and calculate balance based on it

File: spec/models/user_spec.rb [hunk 15]
```diff
@@ -107,6 +108,39 @@
...
+      }
+    end
+
+    context 'when return nil' do
+      it { expect(user.last_payment_date).to eq(nil) }
+    end
+  end
+
+
+  def balance_total
```
Comment (pereslop) [inline spec/models/user_spec.rb:130]: do you need this method here?

**Review by IvanDanyliv — APPROVED** [review]

File: app/models/billing_contract.rb [hunk 16]
```diff
@@ -21,6 +24,12 @@ def self.to_csv
       end
     end
   end
+
+  def balance_utility_provider(bill = last_bill)
+    return 0 unless bill
+
+    payments.created_between(bill.date, bill.next_date).sum(:amount) - bill.amount
```
Comment (chu4) [inline app/models/billing_contract.rb:31]: here its possible that bill amount will negative or positive

Comment (chu4) [reply app/models/billing_contract.rb:31]: some maybe it actually should be a + not a -

**Review by Swooty18 — APPROVED** [review]

**Review by chu4 — APPROVED** [review]
